engine/core/resolver.ts

```diff
@@ -0,0 +1,145 @@
+// deno-lint-ignore-file no-explicit-any
+import {
+  mapObjKeys,
+  PromiseOrValue,
+  Promisified,
+  waitKeys,
+} from "$live/engine/core/utils.ts";
+
+export type ResolveFunc<T = any> = (data: Resolvable<T>) => Promise<T>;
+export interface BaseContext {
+  key: string;
+  resolveId: string;
+  resolve: ResolveFunc;
+}
+
+export type ResolvableObj<T> = {
+  [key in keyof T]: Resolvable<T[key]>;
+};
+export type Resolvable<T = any> = T & { __resolveType: string };
+
+const resolveTypeOf = <T>(
+  resolvable: T | Resolvable<T>,
+): [Omit<T, "__resolveType">, ((p: unknown) => string) | undefined] => {
+  if (isResolvable(resolvable)) {
+    const { __resolveType, ...rest } = resolvable;
+    if (typeof __resolveType === "function") {
+      return [rest as Omit<T, "__resolveType">, __resolveType];
+    }
+    return [rest as Omit<T, "__resolveType">, () => __resolveType];
+  }
+  return [resolvable, undefined];
+};
+
+const isResolvable = <T>(v: T | Resolvable<T>): v is Resolvable<T> => {
+  return (v as { __resolveType: string })?.__resolveType !== undefined;
+};
+
+const isResolvableObj = <T>(
+  v: T | Resolvable<T> | ResolvableObj<T>,
+): v is ResolvableObj<T> => {
+  return !isResolvable(v) && typeof v === "object";
+};
+
+export type AsyncResolver<
+  T = any,
+  TParent = any,
+  TContext extends BaseContext = BaseContext,
+> = (parent: TParent, context: TContext) => Promise<Resolvable<T> | T>;
+
+export type SyncResolver<
+  T = any,
+  TParent = any,
+  TContext extends BaseContext = BaseContext,
+> = (parent: TParent, context: TContext) => Resolvable<T> | T;
+
+export type Resolver<
+  T = any,
+  TParent = any,
+  TContext extends BaseContext = BaseContext,
+> = AsyncResolver<T, TParent, TContext> | SyncResolver<T, TParent, TContext>;
+
+export type ResolverMap<TContext extends BaseContext = BaseContext> = Record<
+  string,
+  Resolver<any, any, TContext>
+>;
+
+async function object<T>(
+  obj: ResolvableObj<T> | T,
+  resolve: <K extends keyof T>(
+    resolvable: Resolvable<T[K]>,
+  ) => PromiseOrValue<T[K]>,
+): Promise<T> {
+  if (obj instanceof Date) {
+    return obj;
```

Comment:
Should we expect that resolving a config will return a serializable object?

Reply:
In that case, this should probably return a number or string

Reply:
Agree, but seems that date is serializable to ISO8601 format by default, so we are fine, right?

<img width="306" alt="image" src="https://user-images.githubusercontent.com/5839364/224827466-3b7812cd-7fd2-4967-9fd7-e10254f55047.png">

Reply:
Yeah, that will work

The only reason I brought it up is that Fresh itself has a limitation of which type of data can be passed as props to an Island, for example: https://fresh.deno.dev/docs/concepts/islands I